### PR TITLE
Add support for enumerations embedded in messages.

### PR DESCRIPTION
**Pull request by robshakir**
Adds support for outputting enumeration leaves that are embedded directly within a message.

```
  * (M) ygen/codegen.go
    - Restructure ordering of messages to use the path of the message
      since the protobuf message name might not be unique, which resulted
      in some messages not being output.
  * (M) ygen/codegen_test.go
    - Add test call for proto generation with enumerations.
  * (M) ygen/protoelements.go
    - Add output for enumerations to be the CamelCase name for the enum
      context leaf.
  * (M) ygen/protogen.go
    - Add template output for protobuf enumerations embedded in
      messages.
  * (M) ygen/protogen_test.go
    - Add test case for enumeration output.
  * (M) ygen/testdata/proto/proto-test-b.compress.formatted-txt
  * (A) ygen/testdata/proto/proto-test-c.device.formatted-txt
  * (A) ygen/testdata/proto/proto-test-c.proto-test-c.elists.elist.formatted-txt
  * (A) ygen/testdata/proto/proto-test-c.proto-test-c.elists.formatted-txt
  * (A) ygen/testdata/proto/proto-test-c.proto-test-c.entity.formatted-txt
  * (A) ygen/testdata/proto/proto-test-c.proto-test-c.formatted-txt
  * (A) ygen/testdata/proto/proto-test-c.yang
    - Add test data for enumerations embedded within a message.
```

### CODE REVIEW

File: ygen/protoelements.go [hunk 1]
```diff
@@ -82,6 +90,15 @@ func (s *genState) yangTypeToProtoScalarType(args resolveTypeArgs) (mappedType,
...
+		// since it will be mapped to the correct name at output file to ensure
+		// that there are no collisions. Enumerations are mapped to an embedded
+		// enum within the message.
+		if args.contextEntry == nil {
+			return mappedType{}, fmt.Errorf("cannot map enumeration without context entry: %v", args)
+		}
+		return mappedType{nativeType: yang.CamelCase(args.contextEntry.Name)}, nil
 	default:
 		// TODO(robjs): implement missing types.
 		//	- enumeration
```
Comment (ostromart) [inline ygen/protoelements.go]: Can remove enumeration from your TODO ;-)

File: ygen/protogen.go [hunk 2]
```diff
@@ -335,7 +349,22 @@ func genProto3Msg(msg *yangDirectory, msgs map[string]*yangDirectory, state *gen
 		case field.IsLeaf() || field.IsLeafList():
 			var protoType mappedType
 			protoType, err = state.yangTypeToProtoType(resolveTypeArgs{yangType: field.Type, contextEntry: field})
-			fieldDef.Type = protoType.nativeType
+			switch {
+			case field.Type.Kind == yang.Yenum && field.Type.Name == "enumeration":
```
Comment (ostromart) [inline ygen/protogen.go]: isSimpleEnum()?

File: ygen/protogen.go [hunk 3]
```diff
@@ -358,6 +387,37 @@ func genProto3Msg(msg *yangDirectory, msgs map[string]*yangDirectory, state *gen
 	return append(msgDefs, msgDef), errs
```
Comment (ostromart) [inline ygen/protogen.go]: Per comments in other review, this function is too long and has too much nesting.

File: ygen/protogen.go [hunk 4]
```diff
@@ -358,6 +387,37 @@ func genProto3Msg(msg *yangDirectory, msgs map[string]*yangDirectory, state *gen
...
 }
 
+// genProtoEnum takes an input yang.Entry that contains an enumerated type
+// and returns a protoMsgEnum that contains its definition within the proto
+// schema.
+func genProtoEnum(field *yang.Entry) (*protoMsgEnum, error) {
+	eval := map[int64]string{}
+	names := field.Type.Enum.NameMap()
+	eval[0] = "UNSET"
```
Comment (ostromart) [inline ygen/protogen.go]: Suggest aggregating proto string literals into a const block somewhere and using those definitions instead.

File: ygen/protogen.go [hunk 5]
```diff
@@ -410,15 +472,32 @@ func genListKeyProto(list *yangDirectory, listName string, listPackage string, s
 			return protoMsg{}, fmt.Errorf("list %s included a key %s that did not have a valid proto type: %v", list.entry.Path(), k, kf.Type)
 		}
 
+		var pt string
+		switch {
+		case kf.Type.Kind == yang.Yenum && kf.Type.Name == "enumeration":
```
Comment (ostromart) [inline ygen/protogen.go]: isSimpleEnum()?

File: ygen/protoelements.go (hunk 1 above)
Comment (robshakir) [reply ygen/protoelements.go]: :+1:, done thanks.

File: ygen/protogen.go (hunk 3 above)
Comment (robshakir) [reply ygen/protogen.go]: Split out leaves in this version too. PTAL.

File: ygen/protogen.go (hunk 5 above)
Comment (robshakir) [reply ygen/protogen.go]: Done, added `isEnumerationLeaf`.

File: ygen/protogen.go (hunk 4 above)
Comment (robshakir) [reply ygen/protogen.go]: Done.

**Comment by robshakir** [comment]
@ostromart -- all comments addressed here, PTAL!

**Pull request merged**